src/transports/smart.c

```diff
@@ -66,6 +66,80 @@ static int git_smart__set_callbacks(
 	return 0;
 }
 
+int http_header_name_length(const char *http_header)
+{
+	const char *colon = strchr(http_header, ':');
+	if (!colon)
+		return 0;
+	return colon - http_header;
+}
+
+bool is_malformed_http_header(const char *http_header)
+{
+	const char *c;
+	int name_len;
+
+	// Disallow \r and \n
+	c = strchr(http_header, '\r');
+	if (c)
+		return true;
+	c = strchr(http_header, '\n');
+	if (c)
+		return true;
+
+	// Require a header name followed by :
+	name_len = http_header_name_length(http_header);
+	if (name_len < 1)
+		return true;
+
+	return false;
+}
+
+static char *forbidden_custom_headers[] = {
+	"User-Agent",
+	"Host",
+	"Accept",
+	"Content-Type",
+	"Transfer-Encoding",
+	"Content-Length",
+};
+
+bool is_forbidden_custom_header(const char *custom_header)
+{
+	unsigned long i;
+	int name_len = http_header_name_length(custom_header);
+
+	// Disallow headers that we set
+	for (i = 0; i < ARRAY_SIZE(forbidden_custom_headers); i++)
+		if (strncmp(forbidden_custom_headers[i], custom_header, name_len) == 0)
+			return true;
+
+	return false;
+}
+
+static int git_smart__set_custom_headers(
+	git_transport *transport,
+	const git_strarray *custom_headers)
+{
+	transport_smart *t = (transport_smart *)transport;
+	size_t i;
+
+	for (i = 0; i < custom_headers->count; i++) {
+		if (is_malformed_http_header(custom_headers->strings[i])) {
+			giterr_set(GITERR_INVALID, "custom HTTP header '%s' is malformed", custom_headers->strings[i]);
+			return -1;
+		}
+		if (is_forbidden_custom_header(custom_headers->strings[i])) {
+			giterr_set(GITERR_INVALID, "custom HTTP header '%s' is already set by libgit2", custom_headers->strings[i]);
+			return -1;
+		}
+	}
+
+	t->custom_headers = custom_headers;
```

Comment:
My general advice is not to re-use remote instances, but we should indeed not crash if someone does it. The `git_remote_fetch()` disconnects from the remote, which means that the second `_fetch()` would have to create a new transport, so the options would not be carried over.

But it is generally good form to defensively copy the array, since this is a setter. While keeping the pointer should be OK in the general case (which is why I didn't comment on it), we could have someone use the lower-level functions (the ones which `git_remote_{fetch,push}()` call), which would introduce a chance for the caller to change the array, or garbage-collect it, or...

So overall, copying the strarray would be good. I don't expect memory usage here to be a concern, so copying a few strings will be fine.